### PR TITLE
Nisdeshmukh/threadpool work item context thandle

### PR DESCRIPTION
**Pull request by nishudeshu**
Changes to make THREADPOOL_WORK_ITEM_CONTEXT as THANDLE. Changed WORK_ITEM to WORK_ITEM_CONTEXT to avoid confusion with individual work_item.

### CODE REVIEW

File: interfaces/inc/c_pal/threadpool.h [hunk 1]
```diff
@@ -22,21 +22,20 @@ extern "C" {
 
 typedef struct THREADPOOL_TAG THREADPOOL;
 typedef struct TIMER_INSTANCE_TAG* TIMER_INSTANCE_HANDLE;
-typedef struct THREADPOOL_WORK_ITEM_TAG* THREADPOOL_WORK_ITEM_HANDLE;
+typedef struct THREADPOOL_WORK_ITEM_CONTEXT_TAG THREADPOOL_WORK_ITEM_CONTEXT;
```
Comment (mattdurak) [inline interfaces/inc/c_pal/threadpool.h:25]: poetry: I think we should remove the word "CONTEXT" here to be more consistent with how we name things (e.g. `THANDLE(THREADPOOL)` is not `THANDLE(THREADPOOL_CONTEXT)`.

Generally "context" is used in our code for some collection of data to pass along with a callback.

File: interfaces/inc/c_pal/threadpool.h [hunk 2]
```diff
@@ -22,21 +22,20 @@ extern "C" {
...
 MOCKABLE_FUNCTION(, int, threadpool_open, THANDLE(THREADPOOL), threadpool);
 MOCKABLE_FUNCTION(, void, threadpool_close, THANDLE(THREADPOOL), threadpool);
 
-MOCKABLE_FUNCTION(, THREADPOOL_WORK_ITEM_HANDLE, threadpool_create_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+MOCKABLE_FUNCTION(, THANDLE(THREADPOOL_WORK_ITEM_CONTEXT), threadpool_create_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
 
-MOCKABLE_FUNCTION(, int, threadpool_schedule_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_ITEM_HANDLE, work_item_context);
-
-MOCKABLE_FUNCTION(, void, threadpool_destroy_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_ITEM_HANDLE, work_item_context);
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work_item, THANDLE(THREADPOOL), threadpool, THANDLE(THREADPOOL_WORK_ITEM_CONTEXT), threadpool_work_item_context);
```
Comment (mattdurak) [inline interfaces/inc/c_pal/threadpool.h:38]: >threadpool_work_item_context [](http://example.com/codeflow?start=128&length=28)

poetry: I also don't think this should say `context` so it should just be `threadpool_work_item`

File: linux/src/threadpool_linux.c [hunk 3]
```diff
@@ -811,25 +797,20 @@ int threadpool_schedule_work_item(THANDLE(THREADPOOL) threadpool, THREADPOOL_WOR
...
 
-                /* Codes_SRS_THREADPOOL_LINUX_05_022: [ threadpool_schedule_work_item shall copy the work_function and work_function_context from threadpool_work_item into insert position in the task array. ]*/
-                task_item->work_function_ctx = threadpool_work_item->work_function_ctx;
-                task_item->work_function = threadpool_work_item->work_function;
+                /* Codes_SRS_THREADPOOL_LINUX_05_022: [ threadpool_schedule_work_item shall copy the work_function and work_function_context from threadpool_work_item_context into insert position in the task array. ]*/
+                task_item->work_function_ctx = threadpool_work_item_context_ptr->work_function_ctx;
+                task_item->work_function = threadpool_work_item_context_ptr->work_function;
 
                 /* Codes_SRS_THREADPOOL_LINUX_05_023: [ threadpool_schedule_work_item shall set the task_state to TASK_WAITING and then release the shared SRW lock by calling srw_lock_release_exclusive. ]*/
                 (void)interlocked_exchange(&task_item->task_state, TASK_WAITING);
```
Comment (mattdurak) [inline linux/src/threadpool_linux.c:810]: >                [](http://example.com/codeflow?start=81&length=15)

nit: consider installing this extension in VS to get rid of trailing spaces [https://marketplace.visualstudio.com/items?itemName=MadsKristensen.TrailingWhitespace64](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.TrailingWhitespace64)

File: linux/tests/threadpool_linux_ut/threadpool_linux_ut.c [hunk 4]
```diff
@@ -1279,15 +1268,12 @@ TEST_FUNCTION(on_timer_callback_does_nothing_with_null_pointer)
 
 TEST_FUNCTION(threadpool_create_work_item_with_NULL_threadpool_fails)
 {
-    // arrange
```
Comment (mattdurak) [inline linux/tests/threadpool_linux_ut/threadpool_linux_ut.c:1282]: >arrange [](http://example.com/codeflow?start=7&length=7)

nit: I think by convention we try to keep the "arrange" comment even if there is nothing that needs to be done under it

File: win32/src/threadpool_win32.c [hunk 5]
```diff
@@ -107,6 +109,16 @@ static void threadpool_dispose(THREADPOOL* threadpool)
     execution_engine_dec_ref(threadpool->execution_engine);
 }
 
+static void threadpool_dispose_work_item(THREADPOOL_WORK_ITEM_CONTEXT_HANDLE threadpool_work_item_context)
+{
+    THREADPOOL_WORK_ITEM_CONTEXT_HANDLE threadpool_work_item_context_ptr = THANDLE_GET_T(THREADPOOL_WORK_ITEM_CONTEXT)(threadpool_work_item_context);
```
Comment (mattdurak) [inline win32/src/threadpool_win32.c:114]: >THANDLE_GET_T [](http://example.com/codeflow?start=75&length=13)

not needed, the dispose function takes a type `T*` for the parameter, not a `THANDLE(T)`. Can remove this line.

File: win32/src/threadpool_win32.c [hunk 6]
```diff
@@ -221,77 +233,56 @@ THREADPOOL_WORK_ITEM_HANDLE threadpool_create_work_item(THANDLE(THREADPOOL) thre
...
+            /* Codes_SRS_THREADPOOL_WIN32_05_009: [ If there are no errors then this threadpool_work_item_context of type THREADPOOL_WORK_ITEM_CONTEXT_HANDLE would be returned indicating a succcess to the caller. ]*/
+            if (threadpool_work_item_context_ptr->ptp_work == NULL)
             {
                 /* Codes_SRS_THREADPOOL_WIN32_05_010: [ If any error occurs, threadpool_create_work_item shall fail, free the newly created context and return a NULL value. ]*/
                 LogError("CreateThreadpoolWork failed");
-                free(work_item_context);
-                work_item_context = NULL;
+                THANDLE_FREE(THREADPOOL_WORK_ITEM_CONTEXT)(threadpool_work_item_context_ptr);
+                threadpool_work_item_context_ptr = NULL;
             }
```
Comment (mattdurak) [inline win32/src/threadpool_win32.c:257]: > } [](http://example.com/codeflow?start=11&length=2)

nit: should have an "else" branch for readability, even if it just has a comment like "ok"

**Comment by mattdurak** [comment]
> #include "ws2tcpip.h"

not your change, but I feel like we shouldn't need these two includes...

---
Refers to: win32/tests/threadpool_win32_ut/threadpool_win32_ut.c:7 in ac801d4. [](commit_id = ac801d4bca3735ba6a789dd43c4d47b419924b1e, deletion_comment = False)

**Review by mattdurak — DISMISSED** [review]
:shipit:

File: interfaces/inc/c_pal/threadpool.h (hunk 1 above)
Comment (nishudeshu) [reply interfaces/inc/c_pal/threadpool.h:25]: Thanks for the context :)

File: interfaces/inc/c_pal/threadpool.h (hunk 2 above)
Comment (nishudeshu) [reply interfaces/inc/c_pal/threadpool.h:38]: Resolved.

File: linux/src/threadpool_linux.c (hunk 3 above)
Comment (nishudeshu) [reply linux/src/threadpool_linux.c:810]: Okay, this one , I am using VS code for Linux, so might have not been installed.

File: linux/tests/threadpool_linux_ut/threadpool_linux_ut.c (hunk 4 above)
Comment (nishudeshu) [reply linux/tests/threadpool_linux_ut/threadpool_linux_ut.c:1282]: Fixed

File: win32/src/threadpool_win32.c (hunk 5 above)
Comment (nishudeshu) [reply win32/src/threadpool_win32.c:114]: Fixed.

File: win32/src/threadpool_win32.c (hunk 6 above)
Comment (nishudeshu) [reply win32/src/threadpool_win32.c:257]: Fixed

**Comment by nishudeshu** [comment]
> #include "ws2tcpip.h"

Fixed

---
In reply to: [2524638376](https://github.com/Azure/c-pal/pull/418#issuecomment-2524638376) [](http://example.com/codeflow?ancestors=2524638376)

---
Refers to: win32/tests/threadpool_win32_ut/threadpool_win32_ut.c:7 in ac801d4. [](commit_id = ac801d4bca3735ba6a789dd43c4d47b419924b1e, deletion_comment = False)

**Review by parth21999 — APPROVED** [review]

**Pull request merged**